Fix reference cycle in our test suite

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106328

In certain cases we capture ErrorMeta in a list. The ErrorMeta objects hold
tracebacks which contain a frame with a local variable that refers to that list.
This change mutates the list on exit from the frame so that it doesn't refer
to the ErrorMeta objects, breaking the cycle.

zdevito committed:
Fix reference cycle in our test suite

In certain cases we capture ErrorMeta in a list. The ErrorMeta objects hold
tracebacks which contain a frame with a local variable that refers to that list.
This change mutates the list on exit from the frame so that it doesn't refer
to the ErrorMeta objects, breaking the cycle.

[ghstack-poisoned]

xwang233 commented:
I'll check this PR with #106111 and #106106 later today. Thanks for the potential fix.

zdevito committed:
Update on "Fix reference cycle in our test suite"


In certain cases we capture ErrorMeta in a list. The ErrorMeta objects hold
tracebacks which contain a frame with a local variable that refers to that list.
This change mutates the list on exit from the frame so that it doesn't refer
to the ErrorMeta objects, breaking the cycle.

[ghstack-poisoned]

zdevito committed:
Update on "Fix reference cycle in our test suite"


In certain cases we capture ErrorMeta in a list. The ErrorMeta objects hold
tracebacks which contain a frame with a local variable that refers to that list.
This change mutates the list on exit from the frame so that it doesn't refer
to the ErrorMeta objects, breaking the cycle.

[ghstack-poisoned]

xwang233 commented:
Hi @zdevito , thank you for the patch. I just tested your PR on our test failures. Test #106106 was fixed but test #106111 still fails. I'll still need to dig deeper on the sparse triangular solve issue.

cc @eqy

zdevito committed:
Update on "Fix reference cycle in our test suite"


In certain cases we capture ErrorMeta in a list. The ErrorMeta objects hold
tracebacks which contain a frame with a local variable that refers to that list.
This change mutates the list on exit from the frame so that it doesn't refer
to the ErrorMeta objects, breaking the cycle.

[ghstack-poisoned]

zdevito committed:
Update on "Fix reference cycle in our test suite"


In certain cases we capture ErrorMeta in a list. The ErrorMeta objects hold
tracebacks which contain a frame with a local variable that refers to that list.
This change mutates the list on exit from the frame so that it doesn't refer
to the ErrorMeta objects, breaking the cycle.

[ghstack-poisoned]

zdevito commented:
@pytorchbot merge
